### ISSUE TITLE
Sign in with Custom (third party) JWT

### ISSUE THREAD

**Issue by henry2man** [supabase/supabase-flutter#479]
**Is your feature request related to a problem? Please describe.**
In our project the authentication must be managed by a third party service. We want to enable RLS and use our own tokens with custom policies in order to fetch data, so we need to use a custom JWT in order to access Supabase.

We was trying to reproduce this article (https://medium.com/@gracew/using-supabase-rls-with-a-custom-auth-provider-b31564172d5d) but we don't find the apropiate API. In a nutshell:

1. Obtain the Supabase JWT Secret: The JWT secret, which is used to sign JWTs, can be found in the Supabase Dashboard under Settings > API > Config. 
2. Store the signed JWT when a user authenticates.
3. Call `supabase.auth.setAuth()` (or Dart/Flutter equivalent API) from the client: When making calls to the Supabase API, it's necessary to call setAuth using the JWT from step 2 to make calls as the user.
4. Use the JWT data: The data from the JWT payload can be used in either your RLS policy or in column default values.

**Describe the solution you'd like**

I want to set my custom session JWT in Supabase, so I can implement custom RLS policies. I want to do it only once (during login process).

Ideally we should have a public method API for setting our custom token into `auth.currentSession.accessToken`, so `_getAuthHeaders()` will take it into consideration (so no header "override" will be needed):

https://github.com/supabase/supabase-flutter/blob/0b7b1c6d0a84047de5dc3f295c3e1c18ff61dd1d/packages/supabase/lib/src/supabase_client.dart#L287-L295

**Describe alternatives you've considered**

~~N/A, the external Auth provider is a requisite. We didn't found an alternative in current API~~

EDIT: @bdlukaa contributed a possible workaround: 

```dart
Supabase.instance.client.rest.setAuth(...).from(...)...
```

EDIT2: HERE is a fully working workaround: https://github.com/supabase/supabase-flutter/issues/479#issuecomment-1559170692

EDIT3: Another possible workaround that can be applied only once until a proper API is added:

```dart
client.headers['Authorization'] = "Bearer $jwt";
```

**Additional context**

https://github.com/supabase/supabase-flutter/issues/479#issuecomment-1559117327
https://medium.com/@gracew/using-supabase-rls-with-a-custom-auth-provider-b31564172d5d
https://github.com/supabase/supabase-js/issues/553
https://github.com/supabase/gotrue/issues/1050
https://supabase.com/docs/guides/realtime/extensions/postgres-changes#custom-tokens

**Comment by henry2man** [comment]
> As a workaround, you can try to do `Supabase.instance.client.rest.setAuth(...).from(...)...`.

@bdlukaa I didn't know you were here too! 😊 Thanks, I'll try it and see if it works...

**Comment by henry2man** [comment]
I've been trying this to work but there is something missing and/or broken.

With this snippet: 

```dart
print(await client.rest.setAuth(
  null,
).from('rls_enabled_table').select())
print(await client.rest.setAuth(
  "xxx.yourcustomjwt.ssss",
).from('rls_enabled_table').select())
```

And this policy:

```sql
CREATE POLICY "Enable read access for authenticated users" ON public.rls_enabled_table FOR PERMISSIVE TO authenticated USING (true) ;
```

I'm sure the custom jwt is valid because Supabase is accepting it (I made a mistake setting the signing key and the service returned a `401` error code). But the source code only returns data if I disable RLS for the table.

**Comment by henry2man** [comment]
Another possible workaround that can be applied only once until a proper API is added:

```dart
client.headers['Authorization'] = "Bearer $jwt";
```

**Comment by henry2man** [comment]
In an internal conversation @bdlukaa recommended to me use `setAuth()` method on each call. 

But I've double checked source code and, actually, I think that setting `headers` in `SupabaseClient` seems a better option because they apply not only for PostgresClient, but also to functions, storage, realtime... (see https://github.com/supabase/supabase-flutter/blob/0b7b1c6d0a84047de5dc3f295c3e1c18ff61dd1d/packages/supabase/lib/src/supabase_client.dart#L73-L102).

If Supabase ensures that setting `headers` will be always an override of any other possible header, this could be the safest method. In this was the case, we may add some clarification / documentation at least in source code (https://github.com/supabase/supabase-flutter/blob/0b7b1c6d0a84047de5dc3f295c3e1c18ff61dd1d/packages/supabase/lib/src/supabase_client.dart#LL72C4-L72C4) or even better in docs.

What do you think?

**Comment by taylorjdawson** [comment]
How would one accomplish this using the javascript application?

**Comment by henry2man** [comment]
> How would one accomplish this using the javascript sdk?

In the original article, the author said this:

https://medium.com/@gracew/using-supabase-rls-with-a-custom-auth-provider-b31564172d5d#:~:text=3.%20Call%20setAuth,Cookies.get(%22mytoken%22))

> _Call `setAuth()` from the client_
> _When making calls to the Supabase API, make sure to call [setAuth](https://supabase.com/docs/reference/javascript/auth-setauth) using the JWT from step 2 to make calls as the user._
> 
> _Using the `js-cookie` package (imported as Cookies):_
> 
> _`supabase.auth.setAuth(Cookies.get("mytoken"))`_

NOTE: in this case a cookie is used to store the token, but this could change in other platforms.

**Comment by henry2man** [comment]
Interesting, it looks that `setAuth()` were removed from current goTrue client implementation:

https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts

But it was present on previous versions: 

https://github.com/supabase/gotrue-js/blob/edcd2ffde6dfe81d02700a52a79e635a5ab8839d/src/GoTrueClient.ts#L389

**Comment by henry2man** [comment]
> Interesting, it looks that `setAuth()` were removed from current goTrue client implementation:

Confirmed: https://github.com/supabase/gotrue-js/pull/340

I have to read the PR well because there was a lot of controversy when removing this method

**Comment by henry2man** [comment]
More on this: 

https://github.com/supabase/supabase-js/issues/553
https://github.com/supabase/gotrue/issues/1050
https://supabase.com/docs/guides/realtime/extensions/postgres-changes#custom-tokens 

TL;DR: Two approaches are discussed: per request custom client vs one time "`signInWithToken`"

Ideally I think we should have a common solution between clients in order to have some coherence between SDKs. @bdlukaa @taylorjdawson can you ask Javascript folks?

We could survive with workarounds...

**Comment by dshukertjr** [comment]
@taylorjdawson It would be great if you could open an issue on supabase-js repo and discuss this there!

**Comment by dshukertjr** [comment]
@henry2man 
Yup, using the heasers setter is the way to go for customer JWT! Note that you are responsible for refreshing the JWT by yourself. 

PR is always welcome if you want to add any comments to the SDK!

**Comment by henry2man** [comment]
> And this policy:
> 
> * table:  public.rls_enabled_table
> * PERMISSIVE TO authenticated
> * USING (true)

@dshukertjr There is one point that is still unresolved.

With a custom token (using headers for now) I'm able to pass Supabase security controls if I pass custom tokens signed with the same secret as Supabase. 

BUT when I try to implement RLS policies, I'm not able to get the `authenticated` role in Postgres. I don't think this is expected in any manner.

Do tokens need any special payload in order get that role? If this is the case we need to add more context and docs in order to use custom authentication.

Please review my previous experiments: 
- https://github.com/supabase/supabase-flutter/issues/479#issuecomment-1559117327
- https://github.com/supabase/supabase-flutter/issues/479#issuecomment-1559170692

**Comment by dshukertjr** [comment]
@henry2man The role comes from the `role` property of the JWT. 

Here is a sample JWT provided by Supabase auth.
```json
{
  "aud": "authenticated",
  "exp": 1615824388,
  "sub": "0334744a-f2a2-4aba-8c8a-6e748f62a172",
  "email": "d.l.solove@gmail.com",
  "app_metadata": {
    "provider": "email"
  },
  "user_metadata": null,
  "role": "authenticated"
}
```

You can read more about it in our auth deep dive guide [here](https://supabase.com/docs/learn/auth-deep-dive/auth-deep-dive-jwts).

**Comment by henry2man** [comment]
@dshukertjr Thanks, this all makes sense now.
